include/spdlog/sinks/eventlog_sink_win32.h

```diff
@@ -0,0 +1,246 @@
+// Distributed under the MIT License (http://opensource.org/licenses/MIT)
+
+namespace eventlog_sink_win32 {
+
+struct win32_error : public spdlog_ex
+{
+    static std::string format(std::string const& func, DWORD error = GetLastError())
+    {
+        return fmt::format(func + " failed ({0})", error);
+    }
+
+    static void report(char const* message) SPDLOG_NOEXCEPT
+    {
+        fprintf(stderr, "%s", message);
+        fflush(stderr);
+    }
+
+    static void report(std::string const& message) SPDLOG_NOEXCEPT
+    {
+        report(message.c_str());
+    }
+
+    win32_error(std::string const& func_name, DWORD error = GetLastError())
+        : spdlog_ex(format(func_name, error), error)
+    {}
+};
+
+struct sid
+{
+    static SID* duplicate(SID* p_src_sid)
+    {
+        if (!IsValidSid(p_src_sid))
+            SPDLOG_THROW(spdlog_ex("sid::duplicate: invalid SID received"));
+
+        DWORD sid_size = ::GetLengthSid(p_src_sid);
+        SID* p_dest_sid = (SID*) ::HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sid_size);
+
+        if (!p_dest_sid)
+            SPDLOG_THROW(win32_error("HeapAlloc"));
+
+        if (!CopySid(sid_size, p_dest_sid, p_src_sid))
+        {
+            sid::free(&p_dest_sid);
+            SPDLOG_THROW(win32_error("CopySid"));
+        }
+
+        return p_dest_sid;
+    }
+
+    static void free(SID** ppsid)
+    {
+        if (ppsid && *ppsid)
+        {
+            if (!HeapFree(GetProcessHeap(), 0, (LPVOID) *ppsid))
+            {
+                SPDLOG_THROW(win32_error("HeapFree"));
+            }
+        }
+
+        *ppsid = nullptr;
+    }
+
+    static SID* get_current_user_sid()
+    {
+        struct process_token_t
+        {
+            HANDLE hToken_;
+            bool hasToken_;
+
+            process_token_t(HANDLE process)
+                : hToken_(0)
+                , hasToken_(OpenProcessToken(process, TOKEN_QUERY, &hToken_))
+            {
+                if (!hasToken_)
+                    SPDLOG_THROW(win32_error("OpenProcessToken"));
+            }
+
+            ~process_token_t()
+            {
+                if (hasToken_ && !CloseHandle(hToken_))
+                    win32_error::report(win32_error::format("CloseHandle"));
+            }
+        } current_process_token(GetCurrentProcess()); // GetCurrentProcess returns pseudohandle, no leak here!
+
+        // Get the required size
+        DWORD tusize = 0;
+        GetTokenInformation(current_process_token.hToken_, TokenUser, NULL, 0, &tusize);
+        if (GetLastError() != ERROR_INSUFFICIENT_BUFFER)
+            SPDLOG_THROW(win32_error("GetTokenInformation"));
+
+        std::vector<unsigned char> buffer(tusize);
+        if (!GetTokenInformation(current_process_token.hToken_, TokenUser, (LPVOID) buffer.data(), tusize, &tusize))
+            SPDLOG_THROW(win32_error("GetTokenInformation"));
+
+        return sid::duplicate((SID *) ((TOKEN_USER*) buffer.data())->User.Sid);
```

Comment:
Do we really need all those complicated sid functions? The "ReportEventA" docs state it can be NULL.

Reply:
Right, but then the user information won't be available in the event viewer, which may sometimes be useful as the event log is a global facility of multi-user environment.